Undefined reference to cv::imread(std::string const&, int)

I am building a project with OpenCV 4 beta and when I use the function cv::imread or cv::imwrite I get an undefined reference to the functions.

My CMakeLists.txt is as follows:

cmake_minimum_required(VERSION 3.1.0)

project(myImageProgram)

set(CMAKE_PREFIX_PATH /home/marios-cellink/libtorch)

find_package(OpenCV 4.0.0 REQUIRED)
find_package(Torch REQUIRED)
add_executable(${PROJECT_NAME} "main.cpp")

message(STATUS "OpenCV library status:")
message(STATUS "    config: ${OPENCV_DIR}")
message(STATUS "    version: ${OpenV_VERSION}")
message(STATUS "    libraries: ${OpenCV_LIBS}")
message(STATUS "    include_path: ${OpenCV_INCLUDE_DIRS}")

message(STATUS "Torch library status:")
message(STATUS "    include_path: ${TORCH_INCLUDE_DIRS}")
message(STATUS "    TORCHLIB: ${TORCH_LIBRARIES}")

target_link_libraries(${PROJECT_NAME} ${OpenCV_LIBS})
target_link_libraries(${PROJECT_NAME} ${TORCH_LIBRARIES})

And the program is as follows:

#include <iostream>
#include <torch/script.h>
#include <opencv2/core.hpp>
#include <opencv2/highgui.hpp>
#include <opencv2/imgproc.hpp>
#include <opencv2/imgcodecs.hpp>

int main()
{
    cv::Mat orig = cv::imread("../myImageProgram/0003_.png");
    cv::Mat img = cv::imread("../myImagePorgram/0003_.png", cv::IMREAD_COLOR);
    std::shared_ptr<torch::jit::script::Module> module = torch::jit::load("../myImageProgram/model.pt");
    return 0;
}


Am I linking OpenCV correctly with cmake? It seems that if I comment out the torch parts from CMake, the program build succesfully. Any ideas?

`target_compile_features(target PUBLIC cxx_std_11)` is supported by CMake 3.8+ only.
Produced CMake configuration files can't be reused by CMake less than CMake 3.8:
```
CMake Error in cpp/CMakeLists.txt:
  Specified unknown feature "cxx_std_11" for target "example_cpp_edge".
```

Current minimal supported CMake version is 3.5.x (Ubuntu 16.04 has 3.5.1 by default).

I faced the same problem

`main.cpp:(.text+0x113): undefined reference to `cv::imread(std::string const&, int)'
main.cpp:(.text+0x179): undefined reference to `cv::imread(std::string const&, int)'
main.cpp:(.text+0x567): undefined reference to `cv::imshow(std::string const&, cv::_InputArray const&)'
main.cpp:(.text+0x856): undefined reference to `cv::imshow(std::string const&, cv::_InputArray const&)'
main.cpp:(.text+0xa17): undefined reference to `cv::imwrite(std::string const&, cv::_InputArray const&, std::vector<int, std::allocator<int> > const&)'
main.cpp:(.text+0xabc): undefined reference to `cv::imshow(std::string const&, cv::_InputArray const&)'
/home/sy/anaconda3/lib/libpng16.so.16: undefined reference to `inflateValidate@ZLIB_1.2.9'
liblbd_mod.so: undefined reference to `cv::write(cv::FileStorage&, std::string const&, int)'
liblbd_mod.so: undefined reference to `cv::FileNode::string() const'
liblbd_mod.so: undefined reference to `cv::Algorithm::getDefaultName() const'
liblbd_mod.so: undefined reference to `cv::error(int, std::string const&, char const*, char const*, int)'
liblbd_mod.so: undefined reference to `cv::operator<<(cv::FileStorage&, std::string const&)'
liblbd_mod.so: undefined reference to `cv::FileStorage::FileStorage(std::string const&, int, std::string const&)'
liblbd_mod.so: undefined reference to `cv::Algorithm::save(std::string const&) const'
collect2: error: ld returned 1 exit status
`
And the relative content of CMakeLists is as follows:
`...
## OpenCV
find_package(OpenCV COMPONENTS core imgproc highgui imgcodecs REQUIRED)
...
target_link_libraries(lbd_mod
        ${OpenCV_LIBRARIES})
target_link_libraries(lbd_mod_test
        lbd_mod
        ${OpenCV_LIBRARIES})
...
target_include_directories(${PYPROJECT_NAME} PUBLIC
        "${CMAKE_CURRENT_SOURCE_DIR}/cpp/include"
        ${Boost_INCLUDE_DIRS}
        ${OpenCV_INCLUDE_DIRS}
        ${PYTHON_INCLUDE_DIRS}
        ${LBD_MOD_INCLUDE}
        )

target_link_libraries(${PYPROJECT_NAME}
        ${Boost_LIBRARIES}
        ${OpenCV_LIBRARIES}
        ${PYTHON_LIBRARIES}
        )
...`

@alalek I also tried u mentioned 

> make VERBOSE=1

It shows
`Linking CXX executable lbd_mod_test
/usr/local/bin/cmake -E cmake_link_script CMakeFiles/lbd_mod_test.dir/link.txt --verbose=1
/usr/bin/c++    -rdynamic CMakeFiles/lbd_mod_test.dir/main.cpp.o  -o lbd_mod_test -Wl,-rpath,/home/sy/code/lbdmod-master/build:/usr/local/lib liblbd_mod.so /usr/local/lib/libopencv_highgui.so.4.1.0 /usr/local/lib/libopencv_videoio.so.4.1.0 /usr/local/lib/libopencv_imgcodecs.so.4.1.0 /usr/local/lib/libopencv_imgproc.so.4.1.0 /usr/local/lib/libopencv_core.so.4.1.0 
CMakeFiles/lbd_mod_test.dir/main.cpp.o: In function `main':
main.cpp:(.text+0x113): undefined reference to `cv::imread(std::string const&, int)'
main.cpp:(.text+0x179): undefined reference to `cv::imread(std::string const&, int)'
main.cpp:(.text+0x567): undefined reference to `cv::imshow(std::string const&, cv::_InputArray const&)'
main.cpp:(.text+0x856): undefined reference to `cv::imshow(std::string const&, cv::_InputArray const&)'
main.cpp:(.text+0xa17): undefined reference to `cv::imwrite(std::string const&, cv::_InputArray const&, std::vector<int, std::allocator<int> > const&)'
main.cpp:(.text+0xabc): undefined reference to `cv::imshow(std::string const&, cv::_InputArray const&)'
/home/sy/anaconda3/lib/libpng16.so.16: undefined reference to `inflateValidate@ZLIB_1.2.9'
liblbd_mod.so: undefined reference to `cv::write(cv::FileStorage&, std::string const&, int)'
liblbd_mod.so: undefined reference to `cv::FileNode::string() const'
liblbd_mod.so: undefined reference to `cv::Algorithm::getDefaultName() const'
liblbd_mod.so: undefined reference to `cv::error(int, std::string const&, char const*, char const*, int)'
liblbd_mod.so: undefined reference to `cv::operator<<(cv::FileStorage&, std::string const&)'
liblbd_mod.so: undefined reference to `cv::FileStorage::FileStorage(std::string const&, int, std::string const&)'
liblbd_mod.so: undefined reference to `cv::Algorithm::save(std::string const&) const'
collect2: error: ld returned 1 exit status
CMakeFiles/lbd_mod_test.dir/build.make:89: recipe for target 'lbd_mod_test' failed
make[2]: *** [lbd_mod_test] Error 1
make[2]: Leaving directory '/home/sy/code/lbdmod-master/build'
CMakeFiles/Makefile2:72: recipe for target 'CMakeFiles/lbd_mod_test.dir/all' failed
make[1]: *** [CMakeFiles/lbd_mod_test.dir/all] Error 2
make[1]: Leaving directory '/home/sy/code/lbdmod-master/build'
Makefile:129: recipe for target 'all' failed
make: *** [all] Error 2
`
So what happened?

I also tried "g++" instead of "make", it turns

> $ g++ main.cpp -L/usr/local/lib -lopencv_calib3d -lopencv_imgproc -lopencv_contrib -lopencv_legacy -o lbd_mod_test
In file included from main.cpp:22:0:
cpp/src/precomp.hpp:54:36: fatal error: opencv2/core/utility.hpp: No such file or directory
 #include "opencv2/core/utility.hpp"

--- a/cmake/templates/OpenCVConfig.cmake.in
+++ b/cmake/templates/OpenCVConfig.cmake.in
@@ -273,6 +273,32 @@ endif()
 # ==============================================================
 set(OpenCV_LIBRARIES ${OpenCV_LIBS})
 
+# Require C++11 features for OpenCV modules
+if(CMAKE_VERSION VERSION_LESS "3.1")
+  if(NOT OpenCV_FIND_QUIETLY AND NOT OPENCV_HIDE_WARNING_COMPILE_FEATURES)
+    message(STATUS "OpenCV: CMake version is low (${CMAKE_VERSION}, required 3.1+). Can't enable C++11 features: https://github.com/opencv/opencv/issues/13000")
+  endif()
+else()
+  set(__target opencv_core)
+  if(TARGET opencv_world)
+    set(__target opencv_world)
+  endif()
+  set(__compile_features cxx_std_11)  # CMake 3.8+
+  if(DEFINED OPENCV_COMPILE_FEATURES)
+    set(__compile_features ${OPENCV_COMPILE_FEATURES})  # custom override
+  elseif(CMAKE_VERSION VERSION_LESS "3.8")
+    set(__compile_features cxx_auto_type cxx_rvalue_references cxx_lambdas)
+  endif()
+  if(__compile_features)
+    # Simulate exported result of target_compile_features(opencv_core PUBLIC ...)
+    set_target_properties(${__target} PROPERTIES
+        INTERFACE_COMPILE_FEATURES "${__compile_features}"
+    )
+  endif()
+  unset(__target)
+  unset(__compile_features)
+endif()
+
 #
 # Some macros for samples
 #

--- a/samples/CMakeLists.example.in
+++ b/samples/CMakeLists.example.in
@@ -12,11 +12,6 @@ file(TO_CMAKE_PATH "${EXAMPLE_FILE}" EXAMPLE_FILE)
 message(STATUS "Project: ${EXAMPLE_NAME}")
 message(STATUS "File   : ${EXAMPLE_FILE}")
 
-# Enable C++11
-set(CMAKE_CXX_STANDARD 11)
-set(CMAKE_CXX_STANDARD_REQUIRED TRUE)
-
-
 # Define project name
 project(${EXAMPLE_NAME})
 

--- a/samples/CMakeLists.txt
+++ b/samples/CMakeLists.txt
@@ -58,10 +58,6 @@ else()
 #===================================================================================================
 cmake_minimum_required(VERSION 3.1)
 
-# Enable C++11
-set(CMAKE_CXX_STANDARD 11)
-set(CMAKE_CXX_STANDARD_REQUIRED TRUE)
-
 project(samples C CXX)
 option(BUILD_EXAMPLES "Build samples" ON)
 

--- a/samples/android/face-detection/jni/CMakeLists.txt
+++ b/samples/android/face-detection/jni/CMakeLists.txt
@@ -1,9 +1,5 @@
 cmake_minimum_required(VERSION 3.6)
 
-# Enable C++11
-set(CMAKE_CXX_STANDARD 11)
-set(CMAKE_CXX_STANDARD_REQUIRED TRUE)
-
 set(target detection_based_tracker)
 project(${target} CXX)
 

--- a/samples/android/tutorial-2-mixedprocessing/jni/CMakeLists.txt
+++ b/samples/android/tutorial-2-mixedprocessing/jni/CMakeLists.txt
@@ -1,9 +1,5 @@
 cmake_minimum_required(VERSION 3.6)
 
-# Enable C++11
-set(CMAKE_CXX_STANDARD 11)
-set(CMAKE_CXX_STANDARD_REQUIRED TRUE)
-
 set(target mixed_sample)
 project(${target} CXX)
 

--- a/samples/android/tutorial-4-opencl/jni/CMakeLists.txt
+++ b/samples/android/tutorial-4-opencl/jni/CMakeLists.txt
@@ -1,9 +1,5 @@
 cmake_minimum_required(VERSION 3.6)
 
-# Enable C++11
-set(CMAKE_CXX_STANDARD 11)
-set(CMAKE_CXX_STANDARD_REQUIRED TRUE)
-
 set(target mixed_sample)
 project(${target} CXX)
 

--- a/samples/cpp/example_cmake/CMakeLists.txt
+++ b/samples/cpp/example_cmake/CMakeLists.txt
@@ -1,10 +1,6 @@
 # cmake needs this line
 cmake_minimum_required(VERSION 3.1)
 
-# Enable C++11
-set(CMAKE_CXX_STANDARD 11)
-set(CMAKE_CXX_STANDARD_REQUIRED TRUE)
-
 # Define project name
 project(opencv_example_project)
 

